### ISSUE TITLE
GraphQL integration

### ISSUE THREAD

**Issue by aradalvand** [edgedb/edgedb#4085]
Hey there. I stumbled into EdgeDB a couple of days ago and so far it seems really promising, I've struggled with a lot of the pain points EdgeDB is meant to solve and so I find the idea behind it brilliant. Hopefully it sees more and more adoption as time goes by.

I have a question about the GraphQL integration. I am aware that EdgeDB provides a built-in GraphQL server through its [GraphQL extension](https://www.edgedb.com/docs/graphql/index) that could be used to run CRUD operations against the database directly via GraphQL. 
Correct me if I'm wrong, but I don't see how this (at least in its current state) is useful if you're actually developing a public-facing API (say, the back-end of a website, for example), as it doesn't allow for any customization, auth, business logic, etc.

When I first heard that EdgeDB plays nicely with GraphQL and before actually reading the docs on the GraphQL integration I was expecting some sort of an engine that would translate GraphQL queries into EdgeQL, something that you could **plug into** frameworks like Apollo Server, the idea of a standalone HTTP server with a GraphQL endpoint seems exotic, and pretty inflexible.

Let me illustrate some very common requirements when building a semi-serious database-driven GraphQL API:

Let's say I have:
```esdl
type Product {
    required property title -> str;
    required property deleted -> bool;
}
```
I don't want any deleted products (i.e. any product whose `deleted` column is `true`) to be exposed via my API.
And then, I will not want the `deleted` column itself to be part of the corresonding GraphQL type either, as it would always be `false` anyway.
GraphQL type fields are usually written in `camcelCase` by convention, while in Postgres and EdgeDB, the convention is `snake_case`. There's currently no way to convert these property names into `camelCase` in the GraphQL API.
Perhaps most importantly, there's no way for us to define any business logic before applying mutations, anybody can delete or edit anything. If this was Apollo, for example, I would've defined a resolver for my mutation and done the various checks there, and the used EdgeDB's Node client to actually make the changes to the database.
There's also no way I can customize exactly which fields the user can do filters on, and much much more.

**Comment by elprans** [comment]
Hi @aradalvand!

Thank you for the feedback!

The bottom line is that the current GraphQL implementation is essentially an experimental proof-of-concept extension.  We should be more upfront about this in the documentation so as not to betray the exectation of top-notch functionality like elsewhere in EdgeDB.  The good news is that support for building secure public facing APIs with nothing but EdgeDB is something we are working toward.  For example, one of the headline features of EdgeDB 2.0 (which is currently in the release candidate status) is the ability to declare precise object-level access control policies, so your "deleted" rule can be expressed as:

```
type Product {
    required property title -> str;
    required property deleted -> bool;
    
    access policy no_deleted allow select using (not .deleted);
}
```

or even:

```
type Product {
    required property title -> str;
    required property deleted -> bool;
    
    access policy no_deleted_in_graphql 
      when (global is_graphql)
      allow select using (not .deleted);
}
```

Control over the exposed API schema as well as explicit support for GraphQL federation is on the roadmap.

A standalone JavaScript library that translates GraphQL to EdgeQL and can be embedded into your app for greatest control is a very nice idea.  GraphQL -> EdgeQL translation is pretty much what the current GraphQL extension does and it's not very hard, since GraphQL is basically a small subset of EdgeQL.

**Comment by aradalvand** [comment]
> A standalone JavaScript library that translates GraphQL to EdgeQL and can be embedded into your app for greatest control is a very nice idea. GraphQL -> EdgeQL translation is pretty much what the current GraphQL extension does and it's not very hard, since GraphQL is basically a small subset of EdgeQL.

Yep, I think that's really the path to take here, and as you said, it shouldn't take too much effort since the current GraphQL extension seems to already do pretty much all of that, so all that's needed is to basically package that as a JavaScript library.

I would even go so far as to say that the focus should probably be on this when it comes to GraphQL integration, as opposed to the standalone GraphQL server thing, as the latter would still be pretty limited even if features like "access control policies" were implemented, providing a lower-level component, like a GraphQL -> EdgeQL transpiler would offer much greater flexibility in addition to eliminating the need for you guys to deal with stuff you shouldn't really have to deal with such as the GraphQL HTTP server configuration and whatnot, so to speak, and just generally seems like a better idea.
Hope you agree.

Thank you.

**Comment by kerimcharfi** [comment]
Hi!
I just started playing with edgedb and I'm very pleased by this project! Keep it going.

> A standalone JavaScript library that translates GraphQL to EdgeQL and can be embedded into your app for greatest control is a very nice idea. GraphQL -> EdgeQL translation is pretty much what the current GraphQL extension does and it's not very hard, since GraphQL is basically a small subset of EdgeQL.

I somewhat disagree. In my opinion the provided crud is a great start and could be used in a many cases. The way I think about it, is to wrap the edgedb graphql with something like graphql-tools with apollo, which allows me to hide/rename/add fields in the schema, do access control, business logic, etc before delegating it to the edgedb database graphql extension. This way I dont have to mess with building filters from scratch and the other nicely provided stuff, because I can build on top of the autogenerated edgedb-graphql schema and can have (in most cases) a very thin resolver in my wrapping server. 

At what point in the application flow could a graphql edgeql translation be useful?  There already exists so much tooling for graphql endpoint wrapping and schema manipulation, parsing, codegen etc., why not tap into it ? 

> Fields in GraphQL are usually written in the camelCase notation, while in Postgres and EdgeDB, the convention is snake_case. There's currently no way to convert these property names into camelCase in the GraphQL API.

In this case a translation vom gql to edgeql would not provide a good solution either.

**Comment by kerimcharfi** [comment]
My question regarding the graphql extension: Can you emit / generate  the graphql schema with the great edgedb cli?
